### DISCUSSION TITLE
Implement print

### DISCUSSION THREAD

**Discussion by llemeurfr** [edrlab/thorium-reader#1737] [Ideas]
It should be possible to print one or more pages inside the app, like we can do in a PDF reader. 
If the publication is LCP protected, the print constraint must be handled properly.

**Comment by llemeurfr** [comment]
Integrating Paged.js may be a solution. In this case it may be better to create a specific window for the selected paginated content. 

In the LCP spec, a page is defined as:
- The page as defined in the Publication, if it is pre-paginated (fixed layout) OR
- The page as defined by the page-list nav element of the EPUB Navigation Document, if this exists OR
- 1024 Unicode characters for all other cases

**Comment by danielweck** [comment]
Looking at other reading systems' implementations, I feel that this is a woefully underspecified domain, i.e. we lack a reliable definition of the problem space. I think that we should gather functional requirements before attempting to implement technical solutions. Assuming there is prior art in the minefield that is "printing from EPUB", we might need to adhere to existing practices. Otherwise, we can just define our own rules, and test them in the field.

* Author intent: possibly expressed with CSS `print` stylesheets, otherwise reliance on the web browser engine's ability to render the markup in a form suitable for printing (e.g. removal of background images / colours, injection of running headers/footers, etc.)
* User intent: print an entire HTML document (aka "chapter"or "spine item" in EPUB parlance), vs. print a set of "pages" from said document (we need a definition of the notion of "page", see bullet point below), vs. print a collated sequence of HTML document (whole publication in linear reading order?), etc.
* What is a "page"? For a reflowable EPUB in scrollview mode, this could be A4 units of print (just like in web browsers, albeit with inconsistent results due to lack of strict specification). For a reflowable EPUB in paginated mode (i.e. CSS columns), this could be a number of visible fragments (i.e. single, or two-page spreads). But, the notion of "pages" can also be dictated by the EPUB3 `nav@epub:type=pagelist` in the Navigation Document (associated with an ordered sequence of `epub:type=page-break` elements scattered in HTML documents). Alternatively, this can be dictated by the reading system's own concept of fixed-size synthetic pages, in Readium2 parlance: "positions list" based on number of "characters" (Unicode code points/units), or number of bytes. For a fixed-layout EPUB, "pages" can intuitively be mapped to individual HTML documents.
* What is the definition of "page" in the LCP DRM? ("rights" limitations imposed by some publishers) => see @llemeurfr 's post immediately above this one.
* etc.

**Comment by danielweck** [comment]
Electron/Chromium "print" API:
https://electronjs.org/docs/api/web-contents#contentsprintoptions-callback

Note this statement: "Use `page-break-before: always;` CSS style to force to print to a new page" (this would require injecting additional markup and CSS at key locations in HTML documents, possibly interfering with the authors / content creators' own intent!)

Some notable configuration parameters for `print()`:

* `printBackground` (Prints the background color and image of the web page. Default is false.)
* `color` (Set whether the printed web page will be in color or grayscale. Default is true.)
* `margins` `marginType` (Can be default, none, printableArea, or custom. If custom is chosen, you will also need to specify `top`, `bottom`, `left`, and `right`.)
* `landscape (Whether the web page should be printed in landscape mode. Default is false.)
* `scaleFactor` (The scale factor of the web page.)
* `pagesPerSheet` (The number of pages to print per page sheet.)
* `collate` (Whether the web page should be collated.)
* `copies` (The number of copies of the web page to print.)
* `pageRanges` (The page range to print. Should have two keys: `from` and `to`.)
* `duplexMode` (Set the duplex mode of the printed web page. Can be simplex, shortEdge, or longEdge.)
* `dpi` (horizontal Number (optional) - The horizontal dpi. vertical Number (optional) - The vertical dpi.)

**Comment by danielweck** [comment]
Interestingly in the Electron/Chromium "print to PDF" API:
https://electronjs.org/docs/api/web-contents#contentsprinttopdfoptions
=> `printSelectionOnly` (which is not available in the regular "print" feature)

**Comment by God-damnit-all** [comment]
@danielweck A print feature would make Thorium Reader the only app in existence, as far as I know of, capable of converting a fixed layout epub3 file to a PDF (via the Save As PDF output option).

If it at all possible, please implement this. It might be a mess for non-fixed layout epubs, but it would be a godsend for fixed layout ones.

**Comment by axu2** [comment]
@God-damnit-all Since then, have you found any other options for fixed layout epub to PDF?
